.github/workflows/ci-pipeline.yml

```diff
@@ -12,7 +12,46 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
+
       - name: Build tests on Docker
         run: docker build --target test --tag todo-app:test .
+
       - name: Run tests
-        run: docker run todo-app:test
\ No newline at end of file
+        run: docker run todo-app:test
+
+  deploy:
+    name: Publish
+    needs: build
+    if: (github.event_name == 'push' && github.ref == 'refs/heads/module-8') # TODO: Change branch to main
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Login to Docker Hub
+        uses: docker/login-action@v2
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Build and push to Docker Hub
+        uses: docker/build-push-action@v3
+        with:
+          push: true
+          target: production
+          tags: framasw/todo-app:latest,framasw/todo-app:${{ github.sha }}
+
+      - name: Login to Heroku
+        env:
+          HEROKU_API_KEY: ${{ secrets.HEROKU_API_KEY }}
```

Comment:
Minor but given you have this twice, it could be worth setting it as a pipe level environment variable rather than a step one - just move exactly this block up to the same level as `runs-on` and `if` etc.